doc/reference/trd-hil-design.md

```diff
@@ -0,0 +1,685 @@
+receive a callback invoked from within the `random_ready` callback.
+  
+Expert programmers who are fully prepared for a re-entrant callback might
+realize this and program accordingly, but most programmers aren't. Some of
+the Tock developers who have been writing event-driven embedded for code
+decades have run into this problem. Having synchronous callbacks makes all
+code need to be as carefully written as interrupt handling code, since
+from the caller's standpoint the callback can preempt execution.
+
+Issuing an asynchronous callback requires that the module be invoked
+again later: it needs to return now, and then after that call stack is
+popped, invoke the callback. For callbacks that will be triggered by
+interrupts, this occurs naturally. However, if, such as in the random
+number generation example, the callback is purely from software, the
+module needs a way to make itself be invoked later, but as quickly as
+possible. The standard mechanism to achieve this in Tock is through
+deferred procedure calls. This mechanism allows a module to tell the
+Tock scheduler to call again later, from the main scheduling loop. For
+example, a caching implementation of `Random` might look like this:
+
+```rust
+impl Random for CachingRNG {
+  fn random(&self) -> Result<(), ErrorCode> {
+    if self.busy.get() {
+      return Err(ErrorCode::BUSY);
+    }
+	
+	self.busy.set(true);
+	if self.cached_words.get() > 0 {
+	  // This tells the scheduler to issue a deferred procedure call,
+	  // passing the "handle" the scheduler uses to keep track of it.
+	  self.handle.map(|handle| self.deferred_call.set(*handle));
+	} else {
+	  self.request_more_randomness();
+	}
+  }
+  ...
+}
+
+impl<'a> DynamicDeferredCallClient for CachingRNG<'a> {
+    fn call(&self, _handle: DeferredCallHandle) {
+	  let rbits = self.pop_cached_word();
+      self.client.random_ready(rbits, Ok(()));
+    }
+}
+
+```
+
+Rule 2: Return Synchronous Errors
+===============================
+
+Methods that invoke hardware can fail. It could be that the hardware is not
+configured as expected, it is powered down, or it has been disabled. Generally
+speaking, every HIL operation should return a Rust `Result` type, whose `Err` 
+variant includes an error code. The Tock kernel provides a standard set of
+error codes, oriented towards system calls, in the `kernel::ErrorCode` enum.
+Sometimes, however, these error codes don't quite fit the use case and so
+a HIL defines its own error codes. The I2C HIL, for example, defines an 
+`i2c::Error` enumeration for cases such as address and data negative
+acknowledgements, which can occur in I2C.
+
+If a method doesn't return a synchronous error, there is no way for a caller
+to know if the operation succeeded. This is especially problematic for
+split-phase calls: whether the operation succeeds indicates whether
+there will be a callback.
+
+Rule 3: Split-phase `Result` Values Indicate Whether a Callback Will Occur
```

Comment:
This section seems incomplete. I assume it was supposed to indicate that `Ok` indicates a callback will occur and `Err` indicates a callback will not.

Reply:
Working on it -- current text is a placeholder, it's not very clear. Will flesh it out tomorrow.

Reply:
Sorry, I didn't mean to push. I forgot this doc is a work in progress.

Reply:
No worries -- I had forgotten I hadn't finished it. :)